# 错误边界
过去，组件内的JavaScript错误会导致React的内部状态被破坏，并且在下一次渲染时产生可能无法追踪的错误。
这些错误基本上是由较早的其他代码（非React组件代码）引起的，但React并没有提供一种在组件中优雅处理这些错误的方式，也无法从错误中恢复。

## 错误边界（Error Boundaries)
部分UI的JavaScript错误不应该导致整个应用崩溃，为了解决这个问题，React 16引入了一个新的概念 ———— 错误边界。

错误边界是一种React组件，这种组件可以捕获并打印发生在其子组件树任何位置的JavaScript错误，并且，他会渲染出备用UI，而不是渲染那些崩溃了的子组件树。错误边界在渲染期间、生命周期方法和整个组件树的构造函数中捕获错误


注意：
错误边界无法捕获以下场景中产生的错误：
1. 事件处理
2. 异步代码（例如setTimeout 或 requestAnimationFrame 回调函数）
3. 服务端渲染
4. 他自身抛出来的错误（并非它的子组件）

如果一个class组件定义了 `static getGervedStateFromError()` 或 `componentDidCatch()`这两个生命周期方法中的任意一个时，那么他就会变成一个错误边界。当抛出错误，请使用`static getDerivedStateFromError()`渲染备用UI，使用`componentDidCatch()`打印错误信息

```jsx
import React from 'react'

export default class ErrorBoundary extends React.Component {
    constructor(props) {
        super(props);
        this.state = { hasError: false };
    }
    static getDerivedStateFromError(error) {
        //更新state使下一次渲染能够显示降级后的UI
        return {hasError: true};
    }
    componentDidCatch(error, errorInfo) {
        //你同样可以将错误日志上报个服务器
        logErrorToMyServic(error, errorInfo);
    } 
    render() {
        if (this.state.hasError) {
            return <h1>Something went wrong.</h1>;
        }
        return this.props.children;
    }
}
```

然后你可以将它作为一个常规组件使用：
```jsx
<ErrorBoundary>
    <MyWidget />
</ErrorBoundary>
```

错误边界的工作方式类似于JavaScript的catch{}, 不同的地方在于错误边界只针对React组件。只有class组件才可以成为错误边界组件。大多数情况下，你只需要声明一次错误边界组件，并在整个应用中使用它。

注意：错误边界仅可以捕获其子组件的错误，它无法捕获其自身的错误。如果一个错误边界无法渲染错误信息，则错误会冒泡至最近的上层错误边界，这也类似于JavaScript中catch{}的工作机制
